Task 3 - MovieDB home page UI with React Components

davidverweij opened the pull request:
This PR (being kept up to date from #2) adds the main layout of the 'NetflixRoulette' home page as per `figma` file. Interactivity is not included. `PropTypes` are generated from the Typescript types using the `typescript-to-proptypes` plugin for Babel.

### Test
- download repo and checkout this branch
- run `npm i` and `npm start` (or `npm run build` for a production build)
- see `PropTypes` in action by changing a `prop` in using the React DevTools. E.g., change a `releaseDate` into a `string`.
- see the `ErrorBoundary` in action by invoking an error at one of the child elements in the React DevTools.

@Dmitry-White, let me know what you think. I wasn't completely sure about the folder structure and have now changed it a few times - some feedback on that is much appreciated! Also, I found it challenging to completely separate 'smart' and 'dumb' components, and was generally finding myself adding a tiny bit of styling on the containers/smart components to align the items within. Any thoughts on that?

Ps. I am assuming that we don't have to create a fully responsive layout - let me know if otherwise!

This branch is currently deployed on gh-pages, see [https://www.davidverweij.com/reactjs-practice-app/](https://www.davidverweij.com/reactjs-practice-app/)

Dmitry-White commented on src/containers/Dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,66 @@
+
+// DATA
+
+const genres = ["All", "Documentary", "Comedy", "Horror", "Crime"];
```
Consider creating a "constants" folder under "src/core" where all the main reusable non-react things will be stored.
"constants" will hold any number of file to your liking - as long as they contain all the hard-coded numbers/strings that you use throughout the app. The idea is to have a central place for all the constants to be able to easily reuse/change them.
Once all the constants are in "src/core/constants" don't forget to go through the app and use them where applicable.

Dmitry-White replied:
It's fine to just import them since it's more of a reuse matter than a design&structure.
Props are usually left for parent-child focused data and/or Dependency Injection things.

Dmitry-White commented on src/containers/Dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,66 @@
+const genres = ["All", "Documentary", "Comedy", "Horror", "Crime"];
+const sortOptions = ["Release date", "Rating", "Runtime", "Title"];
+
+const movies: MovieProps[] = [
```
Under "src/core" folder that you create after one of the above comments consider creating a folder "data" where you would store such arrays/objects to use across the app.

Dmitry-White commented on src/containers/Header/Header.tsx:
```diff
@@ -0,0 +1,20 @@
+  return (
+    <>
+      <img
+        alt="Blurry backdrop displaying video catalog"
```
I can see that you're using a lot of raw user-facing strings which is not extendable and error-prone.
Under "src/core" folder that you create after one of the above comments consider creating a folder "i18y" which stands for "internationalization". This folder in real project usually holds translation object in cases when they're stored in the source code. The folder structure consists of "index.js" file and a number of "*.js" files for different languages (e.g. "en.js" for English or "ge.js" for German). Each of these language files contains an object with the same keys but different values for the target language. The idea behind it is to decouple the selected language for a website and the actual text used in all the components.

E.g. your "Header" will have an image alt text in the JSX as follows:
alt={I18Y['EN'].HEADER_IMAGE_ALT} for English and alt={I18Y['GE'].HEADER_IMAGE_ALT} for German. As you might guess, this would allow to hook up language selection logic from your app that will pass either "EN" or "GE" to the I18Y object, but you the same "HEADER_IMAGE_ALT" key - and end up with different language texts displayed on the live website.

In the end, your task is to create such object (could be only English for now) and go though the whole app and change the hard-coded text to the usage of this I18Y object.

davidverweij replied:
Good one. I've found a few approaches online using third-party libraries, but wanted to implement something more akin to what you described. I've now created a submodule that loads the locales based on input - which should allow a change in language using `useState` (future tasks) and could benefit from lazy loading (next task) instead of the current implementation that loads all locales. What do you think?

davidverweij replied:
@Dmitry-White - In particular, I was thinking to adopt the lazy loading as described [here](https://danlaush.biz/posts/exploring-dynamic-imports#use-a-language-specific-bundle-of-a-component)

Dmitry-White replied:
Completely up to you :)
The primary idea of the comment was to introduce I18Y domain that is present in most of the modern projects.
The way content is stored separately somewhere and the way it's delivered to components differs from tech to tech, however main way we template/interpolate components with dynamic content remains the same.

Dmitry-White commented on src/containers/Dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,66 @@
+    <div className={styles.dashboard}>
+      <FilterBar genres={genres} sortOptions={sortOptions} />
+      <hr className={styles.hr} />
+      <MovieList movies={movies} />
```
Additionally, let's simulate the async action of getting the data here.
Under "src/core" folder that you create after one of the above comments consider creating a folder "api" folder that will hold all the specific API requests in your app.
The files under "src/core/api" could be divided into 2 groups: mocked and real.
Use Promise.resolve fro actual async effect and useEffect to get the data.

davidverweij replied:
I've implemented a mock api method - used a `.ts` file instead of `.json` as I'm not sure yet how image urls will be handled once integrated with the real API. this way, the url's are resolved in the bundle process.

I wasn't sure if you wanted to see a layered approach to the API (e.g., an `index.ts` which uses mocked or real api calls based on a variable/setting) - so let me know what you think.

Dmitry-White reviewed (changes requested):
Looks awesome!
Left a couple of comments for further improvement.
Regarding your question about project structure.
There's actually a number of widely used approaches that could be roughly divided into 2 groups: feature-based and file-based. You can read more on the details in React docs: https://reactjs.org/docs/faq-structure.html
My suggestion in general is to have at least the following structure of your `/src`:
- `/src/core`
- `/src/components`
- `/src/pages`
- `/src/store`
- `/src/utils`
Difference between `/src/components` and `/src/pages` will become clear after we introduce React Router, but high-level - page is a component for every different path link in your app excluding common UI components like Header/Footer.
Difference between `/src/core` and `/src/utils` - utils is purely technical JS stuff like reusable functions, variables etc, whereas core is a business level things like data, api, config constants etc.

Alternatively, you can look into something like Atomic Design (https://atomicdesign.bradfrost.com/chapter-2/)

davidverweij commented on src/components/Menu/Menu.tsx:
```diff
@@ -0,0 +1,29 @@
+      <Logo />
+      <Button
+        className={styles.button}
+        text="+ ADD MOVIE"
```
Good point - I've added an ICON components and used an 'icon library to store and reuse icon SVG throughout the project. Let me know what you think.

davidverweij commented:
@Dmitry-White, thanks for the comments and tips! I've adopted these changes and responded with some detailed thoughts on each change request individually. Let me know what you think!

Dmitry-White commented on src/containers/MovieList/MovieList.tsx:
```diff
@@ -0,0 +1,40 @@
+  return (
+    <>
+      <div className={styles.stats}>
+        <span className={styles.thick}>39</span> {I18Y("EN").MOVIES_FOUND}
```
Consider making "39" dynamic, i.e length of the effective movie list.

Dmitry-White commented on src/containers/Dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,31 @@
+const Dashboard = (): JSX.Element => {
+  const [moviesState, setMoviesState] = useState([]);
+
+  const fetchMovies = async (): Promise<void> => {
```
It would make sense to move "fetchMovies" to "core/api".
"fetchMovies" looks like a good abstraction on top of real/mocked api, so that components don't really know/care about the data source.
Therefore, "setMoviesState" also should be moved, e.g to the "useEffect" hook.

Dmitry-White commented on src/core/i18y/index.ts:
```diff
@@ -0,0 +1,18 @@
+
+const I18Y = (language: AvailableLanguages): I18yStrings => {
+  switch (language) {
+    case "NL":
```
In the comment response you mentioned "useState" - I think it would be great to actually have a app-level toggle for languages  switching state :)

Dmitry-White reviewed (changes requested):
Nice work with the I18Y!

davidverweij commented:
Thanks @Dmitry-White! I implemented the changes and got a bit carried away and implemented the language toggle using `useContext` - let me know what you think of this approach. It's also deployed at https://www.davidverweij.com/reactjs-practice-app/

Dmitry-White reviewed (approved):
Wow, it's actually really awesome that you're already familiar with React Context!
Approving the PR, though do make sure to go through the app and finish the translating all the UI components (things like Release Date, or Drama, or Documentaries). You'll need a bit of remapping for these things since you now keep them in "constants" - create the remapping in the same "constants" directory. Hopefully, now it's becoming a bit more obvious why these "constants" are kept separately and how easy it is to change their values and not break the app :)

davidverweij commented:
Thanks @Dmitry-White! I've now also added locales/translation for the constants in a similar manner to I18Y (by using context which affects which strings are returned).
